ui/pages/confirmations/hooks/useCurrentConfirmation.ts

```diff
@@ -1,95 +1,85 @@
-      if (paramsTransactionId === currentConfirmation?.id) {
-        return;
-      }
-      pendingConfirmation = pendingConfirmations.find(
-        ({ id: confirmId }) => confirmId === paramsTransactionId,
-      );
-    }
+  const transactionMetadata = useSelector((state) =>
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (getUnapprovedTransaction as any)(state, confirmationId),
+  ) as TransactionMeta | undefined;
 
-    if (!pendingConfirmation) {
-      if (!latestPendingConfirmation) {
-        setCurrentConfirmation(undefined);
-        return;
-      }
-      pendingConfirmation = latestPendingConfirmation;
-    }
+  const signatureMessage = useSelector((state) =>
+    selectUnapprovedMessage(state, confirmationId),
+  );
 
-    if (pendingConfirmation.id !== currentConfirmation?.id) {
-      const unconfirmedTransaction =
-        unconfirmedTransactions[pendingConfirmation.id];
-      if (!unconfirmedTransactions) {
-        setCurrentConfirmation(undefined);
-        return;
-      }
+  const isCorrectTransactionType = REDESIGN_TRANSACTION_TYPES.includes(
+    transactionMetadata?.type as TransactionType,
+  );
 
-      const isConfirmationRedesignType =
-        REDESIGN_APPROVAL_TYPES.find(
-          (type) => type === pendingConfirmation?.type,
-        ) ||
-        REDESIGN_TRANSACTION_TYPES.find(
-          (type) => type === unconfirmedTransaction?.type,
-        );
+  const isCorrectApprovalType = REDESIGN_APPROVAL_TYPES.includes(
+    pendingApproval?.type as ApprovalType,
+  );
 
-      if (!isConfirmationRedesignType) {
-        setCurrentConfirmation(undefined);
-        return;
-      }
+  const isSIWE =
+    pendingApproval?.type === TransactionType.personalSign &&
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (signatureMessage?.msgParams as any)?.siwe?.isSIWEMessage;
 
-      // comment if condition below to enable re-design for SIWE signatures
-      // this can be removed once SIWE code changes are completed
-      if (pendingConfirmation?.type === ApprovalType.PersonalSign) {
-        const { siwe } = unconfirmedTransaction.msgParams;
-        if (siwe?.isSIWEMessage) {
-          setCurrentConfirmation(undefined);
-          return;
-        }
-      }
-      setCurrentConfirmation(unconfirmedTransaction);
+  return useMemo(() => {
+    if (
+      !redesignedConfirmationsEnabled ||
+      (!isCorrectTransactionType && !isCorrectApprovalType) ||
+      isSIWE
+    ) {
+      return { currentConfirmation: undefined };
     }
-  }, [latestPendingConfirmation, paramsTransactionId, unconfirmedTransactions]);
 
-  return { currentConfirmation };
+    const currentConfirmation =
+      transactionMetadata ?? signatureMessage ?? undefined;
+
+    return { currentConfirmation };
+  }, [
+    redesignedConfirmationsEnabled,
+    isCorrectTransactionType,
+    isCorrectApprovalType,
+    isSIWE,
+    transactionMetadata,
+    signatureMessage,
+  ]);
```

Comment:
That's correct, the performance cost of any logic here is negligible so the purpose of the `useMemo` is to only return a fresh object reference when the underlying data has changed, to prevent all the dependent components re-rendering.

Any iteration over the controller state is encapsulated within the selectors, which is a better separation of concerns, plus we can benefit from deep equal selectors that also only return new object references when the values themselves have changed.